cyto_dl/models/jepa/jepa_base.py

```diff
@@ -0,0 +1,116 @@
+        x_key : str
+            The key used to access the input data.
+        momentum : float, optional
+            The momentum value for the exponential moving average of the model weights (default is 0.998).
+        max_epochs : int, optional
+            The maximum number of training epochs (default is 100).
+        **base_kwargs : dict
+            Additional arguments passed to the BaseModel.
+        """
+        _DEFAULT_METRICS = {
+            "train/loss": MeanMetric(),
+            "val/loss": MeanMetric(),
+            "test/loss": MeanMetric(),
+        }
+        metrics = base_kwargs.pop("metrics", _DEFAULT_METRICS)
+        super().__init__(metrics=metrics, **base_kwargs)
+
+        self.encoder = encoder
+        self.predictor = predictor
+
+        self.teacher = copy.deepcopy(self.encoder)
+        for p in self.teacher.parameters():
+            p.requires_grad = False
+
+        self.loss = torch.nn.L1Loss()
+
+    def configure_optimizers(self):
+        optimizer = self.optimizer(
+            list(self.encoder.parameters()) + list(self.predictor.parameters()),
+        )
+        scheduler = self.lr_scheduler(optimizer=optimizer)
+        return {
+            "optimizer": optimizer,
+            "lr_scheduler": {
+                "scheduler": scheduler,
+                "monitor": "val/loss",
+                "frequency": 1,
+            },
+        }
+
+    def forward(self, x):
+        return self.encoder(x)
+
+    def _get_momentum(self):
+        # linearly increase the momentum from self.momentum to 1 over course of self.hparam.max_epochs
+        return (
+            self.hparams.momentum
+            + (1 - self.hparams.momentum) * self.current_epoch / self.hparams.max_epochs
+        )
+
+    def update_teacher(self):
+        # ema of teacher
+        momentum = self._get_momentum()
+        # momentum update of the parameters of the teacher network
+        with torch.no_grad():
+            for param_q, param_k in zip(self.encoder.parameters(), self.teacher.parameters()):
+                param_k.data.mul_(momentum).add_((1 - momentum) * param_q.detach().data)
+
+    def get_target_embeddings(self, x, mask):
+        # embed the target with full context for maximally informative embeddings
+        with torch.no_grad():
+            target_embeddings = self.teacher(x)
+        target_embeddings = rearrange(target_embeddings, "b t c -> t b c")
+        target_embeddings = take_indexes(target_embeddings, mask)
+        target_embeddings = rearrange(target_embeddings, "t b c -> b t c")
+        return target_embeddings
+
+    def get_context_embeddings(self, x, mask):
+        # mask context pre-embedding to prevent leakage of target information
+        context_patches, _, _, _ = self.encoder.patchify(x, 0)
+        context_patches = take_indexes(context_patches, mask)
+        context_patches = self.encoder.transformer_forward(context_patches)
```

Comment:
can you replace transformer_foward here with regular forward for simplicity? the only difference here is applying the mask to the context patches right?

Reply:
as it's written now, using regular forward would leak target information into the context. Instead of having transformer_forward and forward as separate functions, I could just add a `patchify` flag to the forward pass and set it to false here?

Reply:
this makes sense